[jinyuan0425] iP

jinyuan0425 opened the pull request:
# ButlerBot

## Features
This bot aims to help you keep track of your taskings through:
- Adding of tasks (Todo, Deadline & Event)
- Deleting of tasks
- Marking and unmarking completion of specific task
- Filtering the list of task using a specific keyword
- Listing all tasks in your list

## Download
To get this bot, simply follow these steps:
1. Download it from [here](https://github.com/jinyuan0425/ip/releases/tag/v1.0)
2. Double click to run it
3. Add your tasks
4. Watch as it manages your tasks for you :sunglasses:

Alternatively you may open up your command line and run the command `java -jar {filename}.jar`, or enter the following code in java:

```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

## Commands
The following commands are available for the bot:
- [ ] **list**: __displays your task list__
- [ ] **todo**: __adds a task to your list__
- [ ] **deadline**: __adds a task with a deadline to your list__
- [ ] **event**: __add a task with a start and end date to your list__
- [ ] **filter**: __displays your list of tasks containing the keyword__
- [ ] **delete**: __deletes a specific task off your list__
- [ ] **mark**: __marks a specific task on your list as completed__
- [ ] **unmark**: unmark a specific completed task
- [x] __And many more to come...__

## Reviews
Here are some totally legit reviews by users:
> Best bot ever!
> This bot changed my life.

jinyuan0425 committed:
Completed Level-0

jinyuan0425 committed:
Completed Level-1

jinyuan0425 committed:
Completed Level-2

jinyuan0425 committed:
Completed Level-3

jinyuan0425 committed:
Completed Level-4

jinyuan0425 committed:
Completed A-TextUiTesting

jinyuan0425 committed:
Completed Level-5

Implemented OOP

jinyuan0425 committed:
Completed Level-6

jinyuan0425 committed:
Completed A-Enums

jinyuan0425 committed:
Update code to allow for save

jinyuan0425 committed:
Update code to allow for save

Bot refreshers with each load.

Users will not be able to use the bot effectively as data has to be
re-entered every load.

Let's save the user's input when assigning tasks to the bot.

Bot will now be able to reference the saved inputs and load the user's
saved data.

jinyuan0425 committed:
Merge branch 'branch-Level-7'

jinyuan0425 committed:
Update bot to understand dates

jinyuan0425 committed:
Update bot to understand date and time

There is no fixed format to the input for by, from and to. Bot is able
to accept any input and store them as a valid date and time.

Let's introduce LocalDateTime to the bot to enforce a specific input
format for the date and time input.

This minimises the random behaviour of the bot as any incorrect inputs
will be rejected by the bot.

jinyuan0425 committed:
Merge branch 'branch-Level-8'

jinyuan0425 committed:
Update code to be more OOP

The code works as intended.

Code can be further extracted to improve OOP.

Extracting the common behavior into a super class allows us to use
polymorphism when dealing with command attributes.

Let's extract and group similar commands in to various classes that
extends the parent class Command.

This improves readability and allows for inheritance as well.

jinyuan0425 committed:
Update code to implement packages

Code was organised due to OOP, but can be further organised into
packages.

Let's group the classes into various functionalities and sort them
into the corresponding packages.

This avoids name conflict and helps with writing a better maintainable
code.

jinyuan0425 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

jinyuan0425 committed:
Update code to include testing

Code is working for manual inputs.

Let's use JUnit to automate the test cases.

This allows for a quicker check of the code when making small tweaks
and not have the need to manually input the same inputs each time.

jinyuan0425 committed:
Package the app as an executable file

Bot is now a working prototype.

Let's package it into a JAR file to be executed.

Remote access of the app is now available with this executable file.

jinyuan0425 committed:
Add Javadoc comments to methods

Most methods, less the getters and setters, classes and fields now have
Javadocs comments to detail the function of said methods, class and
fields.

jinyuan0425 committed:
Update code to follow coding standard

Let's update the code to follow the java coding standard.

jinyuan0425 committed:
Update code to allow for find command

Let's implement a find command to filter through the task list and
return a new list of task containg the keyword.

Several test cases are added for more extensive testing.

jinyuan0425 committed:
Merge branch 'branch-A-CodingStandard'

jinyuan0425 committed:
Merge branch 'branch-Level-9'

jinyuan0425 committed:
Update code to use checkstyle

Code will now use checkStyle to detect coding style violations.

jinyuan0425 committed:
Update chatbot to contain GUI

Let's use JavaFX to implement GUI for the chat bot.

WeeMingQing commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,127 @@
+import duke.tasks.Deadline;
+import duke.tasks.Event;
+import duke.tasks.Task;
+import duke.tasks.Todo;
```
Well Done on listing all the classes explicitly and in alphabetical order, and in other places as well.

WeeMingQing commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,127 @@
+    private LocalDateTime deadline;
+    private LocalDateTime from;
+    private LocalDateTime to;
+    private boolean isExit = false;
```
Good job in using the proper naming convention for boolean variables, and in other places as well.

WeeMingQing commented on src/main/java/duke/command/FilterCommand.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * keyword contains the keyword to filter.
+     */
+    private String type;
```
Well Done on declaring class variables as not public.

WeeMingQing reviewed (commented):
LGTM. Well Done.

nabonitasen commented on src/main/java/duke/components/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Exception class for the bot.
+ */
+public class DukeException extends Exception {
+    /**
```
Would it help to categorise this into different subclasses of DukeExceptions?

nabonitasen commented on src/main/java/duke/components/Storage.java:
```diff
@@ -0,0 +1,125 @@
+        }
+    }
+
+    /**
+     * Reads the data stored in the save file and return the corresponding Task.
+     *
+     * @param taskData Data for Task stored in the save file.
+     * @return Task representation of the String stored.
+     * @throws DukeException
+     */
+    private Task parseTask(String taskData) throws DukeException {
```
Good job on going the extra mile in adding comments for private functions!

nabonitasen commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,65 @@
+        this.by = by;
+    }
+
+    /**
+     * Returns the String representation of the Deadline, along with the
+     * indication of whether it isDone and the by of the Deadline.
```
I like that your comments are straightforward and clear!

nabonitasen reviewed (commented):
Looks really good overall! One small change you could consider adding.

jinyuan0425 committed:
Update code to generate JAR file

Let's update the build.gradle file to generate the JAR file for the
bot.

jinyuan0425 committed:
Update code to include assertions

Code has exceptions to handle any unsual conditions created by the
user. However, there is a lack of handlers for errors by the
programmer.

This makes the code prone to bugs due to mistakes by the programmer.
Without a handler for these potential bugs, the bot might break while
running.

Let's include assertions to account for mistakes by the programmer.

This makes the code safe by terminating the execution when there is
a possible bug detected.

jinyuan0425 committed:
Improve code quality

Code, while functional, lacks readability and is error prone. This
increases the likelihood of having bugs that are harder to detect.

Let's improve the code quality by refactoring the code.

This makes the code more readable and less prone to errors. Even if
errors are detected, it will be easier to locate the bugs as well.

jinyuan0425 committed:
Merge pull request #1 from jinyuan0425/branch-A-Assertions

Update code to include assertions

jinyuan0425 committed:
Merge branch 'master' into branch-A-CodeQuality

jinyuan0425 committed:
Merge pull request #2 from jinyuan0425/branch-A-CodeQuality

Improve code quality

jinyuan0425 committed:
Update code to perform continuous integration

Let's include a .yml file such that GitHub Actions will run the
workflow (as per the .yml file) every time certain project events are
triggered (e.g. when a PR is updated, or the master branch is updated).

jinyuan0425 committed:
Merge pull request #3 from jinyuan0425/branch-A-CI

Update code to perform continuous integration

jinyuan0425 committed:
Update code to include tags

Currently, users are unable to organise their tasks in any way.

Let's include tags to allow for better organisation of tasks.

This can further be improved with tag filters, tag sorting etc.
With this, users can have a better view of their to-do list, i.e.
tagging the type of activity, the priority etc.

jinyuan0425 committed:
Merge pull request #4 from jinyuan0425/branch-C-Tagging

Update code to include tags

jinyuan0425 committed:
Rename folder from workflow to workflows

jinyuan0425 committed:
Update code to produce better GUI

Bot dialog box now has a different colour and auto fits text length.

jinyuan0425 committed:
Update header comments

jinyuan0425 committed:
Update user guide

jinyuan0425 committed:
Improve testing

jinyuan0425 committed:
Update bot for v2.0 release
